keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,533 @@
+// on specified policies. HookAction will be created by the user and
+// can be re-used later. Snapshot Controller will create ExecutionHooks
+// based on HookActions specified in the snapshot spec. For example,
+// two HookActions, preSnapshotExecutionHook and postSnapshotExecutionHook,
+// are expected in the snapshot spec.
+// HookAction does not contain information on pods/containers because those are
+// runtime info.
+// HookAction is namespaced
+type HookAction struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // This contains the command to run on a container.
+	// The command should be idempotent because the system does not guarantee exactly-once semantics.
+	// Any action may be triggered more than once but only the latest results will be logged in status.
+	// As alpha feature, only ExecAction type in Handler will be support, not the HTTPGETAction or TCPSocketAction.
+        // This is required.
+        Action core_v1.Handler
+
+        // ActionTimeoutSeconds defines when the execution hook controller should stop retrying.
+        // If execution fails, the execution hook controller will keep retrying until reaching
+        // ActionTimeoutSeconds. If execution still fails or hangs, execution hook controller
+        // stops retrying and updates executionhook status to failed.
+        // +optional
+        ActionTimeoutSeconds *int64
+}
+```
+
+`ExecutionHook` may be used for different cases, such as
+
+* Application-consistency snapshotting
+* Upgrade
+* Rolling upgrade
+* Debugging
+* Prepare for some lifecycle event like a database migration
+* Reload a config file
+* Restart a container
+
+
+The following gives an example of how to use ExecutionHook for application-consistency snapshotting use case.
+
+### Workflow
+
+* Snapshot API should carry two execution hook information (which will reference two HookAction - preSnapshotExecutionHook and postSnapshotExecutionHook) for quiescing and unquiescing application.
+* Snapshot controller watches Snapshot objects. If there is a request to create a Snapshot, it checks if there is ExecutionHook Information defined in Snapshot.
+* If hook is defined in the snapshot, snapshot controller will create one or multiple ExecutionHooks for quiescing application if necessary, one for each command running in pods/containers.
+* Snapshot controller waits for all Action hooks to complete running before taking snapshot.
+* The ExecutionHook controller watches the ExecutionHook API object and take actions based on the object status and also update the status. 
+* Snapshot controller waits until hook is run on all pods (if more than 1 pod). No matter the Action succeeds or fails, snapshot controller should create execution hooks for unquiescing application. Snapshot controller can also decide when to delete those hooks.
+
+Here is an example of an HookAction:
+```
+apiVersion: apps.k8s.io/v1alpha1
```

Comment:
Is `apps.k8s.io` the group name we will use?

Reply:
Yes

Reply:
yeah, we compared a few groups, apps seems the most appropriate one since it is very relevant to application execution.